keps/sig-api-machinery/4346-informer-metrics/README.md

```diff
@@ -0,0 +1,973 @@
+This enhancement proposal adds metrics to the client-go informer. It will expose reflector/queue/eventHandler internal metrics to Prometheus. These metrics is useful for developers/reliability engineers, they can monitor informer depend on it.
+
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+A Kubernetes controller will watch objects for the desired state and the actual state, then send instructions to make the actual state be more like the desired state. Most controllers use informer to watch object change, then send work items that require reconcile to the `workqueue`.
+
+Now the workqueue exposes metrics about queueLatency/workDuration, it is useful to find issues in reconcile routine.  When a lot of objects need to be reconciled, but there are no new work items sent into `workqueue`, the informer most likely blocked. Informer is composed of reflector/queue/eventHandler, to find the root cause, users have to add debug log and change log level.
+
+Informer should expose reflector/queue/eventHandler metrics, it will be easy to find why this informer is blocked. For example, it will show how long in seconds eventHandler processing an item.
+
+This change remove reflector metrics before https://github.com/kubernetes/kubernetes/pull/74636. It is essential to fix memory leak issue.
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Add metrics for informer 
+- Expose informer reflector/queue/eventHandler metrics
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- It does not introduce breaking changes for controllers which use informer.
+- It does not modify core Kubernetes components which use informer.
+- It does not list all informer metrics, which can add as needed
+
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+- Introduce the informer metrics struct `informerMetrics` contains queue/eventHandler metrics
+- Introduce the informer metrics provider interface `informerMetricsProvider`, implement in `k8s.io/component-base/metrics`
+- Add an environment to enable informer metrics
```

Comment:
why do we need an environment variable to enable them?

Reply:
> why do we need an environment variable to enable them?

I think this got updated to a featuregate.  This type of metric has been attempted at least once (maybe twice?) and reverted for problems each time.  Being able to disable our first attempt could help avoid a rush to patch all the impacted clients.

Reply:
Replace it with featuregate, PTAL

Reply:
> Replace it with featuregate, PTAL

forgot to push?